Feature/hooks: Add more support for hooks

nstolpe opened the pull request:
Most of the changes in this PR are originally from [PR-79](https://github.com/michalochman/react-pixi-fiber/pull/79). The objective is the same, adding support for hooks in react-pixi-fiber and adding some built in hooks to use.

nstolpe commented on examples/src/HooksExample/DraggableContainer.js:
```diff
@@ -0,0 +1,40 @@
+import { CustomPIXIComponent } from "react-pixi-fiber";
```
Need to remove this file.

nstolpe commented on examples/src/HooksExample/index.js:
```diff
@@ -0,0 +1,50 @@
+import React, { Component } from "react";
```
This example needs to actually use a hook. Or be removed.

michalochman replied:
Better to leave it and use hooks

michalochman commented:
@nstolpe there are 22 tests that are failing now, and it would be good to have test coverage of hooks as well.

nstolpe commented:
@michalochman I can fix those. I don't think I copied the updated tests over from PR 79 when I got everything else. I was waiting to from you about [this comment](https://github.com/michalochman/react-pixi-fiber/pull/79#issuecomment-509371901) before getting to them though. Since function components and hooks are working in react-pixi-fiber components with the dependency updates you made and the downgrade of scheduler to 0.13.6 (so function components/hooks are working without needing the logic in `Stage.js` that returns it as a class or function depending on React's capabilities), did you want to keep those changes to `Stage.js` in the PR or just keep the `hooks.js` and other associated changes?

michalochman commented:
@nstolpe as long as `Stage` component works the same way from user perspective we shouldn't care whether it's implemented as is or using hooks. :)

michalochman commented on src/Stage.js:
```diff
@@ -61,11 +61,14 @@ export function createStageFunction() {
 
 export function createStageClass() {
   class Stage extends React.Component {
+    static propTypes = propTypes;
```
Is this necessary? I would like to avoid using syntax, especially that few lines above the function component is using conventional `Stage.propTypes = propTypes;` syntax

nstolpe replied:
Yeah, sorry, meant to leave a comment with this too. The proposal seems a little weird to me too. I added this static property and the dependency to get rid of some errors in the linter (only once it was pushed here though, locally it was fine), all of the stage props were getting this error: `<prop> is missing in props validation`.

I'll revert it back.

nstolpe replied:
I was able to get the eslint issues locally, all when accessing props in the class component:

```
  70:15  error  'children' is missing in props validation  react/prop-types
  70:25  error  'height' is missing in props validation    react/prop-types
  70:33  error  'options' is missing in props validation   react/prop-types
  70:42  error  'width' is missing in props validation     react/prop-types
  80:15  error  'children' is missing in props validation  react/prop-types
  80:25  error  'height' is missing in props validation    react/prop-types
  80:33  error  'options' is missing in props validation   react/prop-types
  80:42  error  'width' is missing in props validation     react/prop-types
  81:15  error  'options' is missing in props validation   react/prop-types
  92:15  error  'options' is missing in props validation   react/prop-types
```

Do you know what setting could be causing those?

michalochman replied:
Does it happen also when you do `Stage.propTypes = propTypes` when you define `Stage` in either `createStageClass` or `createStageFunction` instead of at the end of the file?

nstolpe replied:
I got it to pass by assigning `Stage.propTypes` in both of the `create`  functions instead of at the end.

nstolpe commented:
@michalochman I got the test coverage up and the linting fixed. I'll try and get the example finished tomorrow. Let me know if you have any feedback on the tests.

nstolpe commented:
Hooks example updated and working now too.

michalochman commented on src/hooks.js:
```diff
@@ -0,0 +1,61 @@
+export function usePixiApp(props) {
+  const { options, width, height } = props;
+  const canvasRef = useRef(null);
+  const [app, setApp] = useState(null);
```
It doesn't look like the state is actually used anywhere.

nstolpe replied:
It's part of the return object, along with canvas.

michalochman replied:
Right, the application is created in an effect.

michalochman commented on src/hooks.js:
```diff
@@ -0,0 +1,61 @@
+      unmount(appInstance.stage);
+      appInstance.destroy();
+    };
+  }, []);
```
I think that the dependency list should not be empty – now it's impossible to change `<Stage />` props like options (for example backgroundColor).

michalochman commented on src/hooks.js:
```diff
@@ -0,0 +1,61 @@
+import { getCanvasProps } from "./stageProps";
+import { unmount } from "./render";
+
+export function usePixi() {
```
Also, this should be re-exported in main index file, so users can actually import it. :)

michalochman commented on src/hooks.js:
```diff
@@ -0,0 +1,61 @@
+  return useContext(AppContext);
+}
+
+export function usePixiTicker(fn) {
```
Please re-export in index file.

nstolpe commented on src/Stage.js:
```diff
@@ -1,116 +1,113 @@
-    const stageProps = getDisplayObjectProps(this.props);
-    applyProps(this._app.stage, {}, stageProps);
+      applyUpdate(app, props);
+      resizeRenderer(app, prevProps, props);
```
Looks like the original code was doing the logic in applyUpdate first, I'll modify the class to use this order.

nstolpe commented on src/stageProps.js:
```diff
@@ -0,0 +1,51 @@
+import PropTypes from "prop-types";
```
None that I can tell. I've moved `stageProps` back into `Stage`.

michalochman replied:
We now know that the reason was circular dependencies issue

nstolpe commented on test/StageLegacy.test.js:
```diff
@@ -0,0 +1,336 @@
+import * as PIXI from "pixi.js";
+import { Text } from "../src/index";
+import ReactPixiFiber from "../src/ReactPixiFiber";
+import Stage from "../src/Stage";
```
I thought I'd used the `create` function here before, must have reverted it. I've updated to use it now, I'll combine the tests too.

nstolpe replied:
Used the `createStage...` functions and combined the tests.

nstolpe commented:
Thanks @michalochman. I've addressed most of your comments. There are a few more I'll get to tomorrow or over the weekend.

michalochman commented on src/index.js:
```diff
@@ -3,11 +3,25 @@ import { AppContext, AppProvider, withApp } from "./AppProvider";
+  unstable_batchedUpdates,
+  usePixiApp,
+  usePixiTicker,
+  usePreviousProps,
```
This is mostly for internal use, I think we should not export it. Even though it's simple and useful it is not related to the API that is already surfaced to users.

michalochman commented:
@nstolpe let me know if you need any help. I've also did an update that you might want to inspect in #132 that will require a rebase here.

nstolpe commented:
@michalochman I may need some help w/ the dependencies for useLayoutEffect...that breaks the tests and framework whenever I've added any. I should have some time tonight to get to this and at least get the rebase conflicts fixed.

michalochman commented:
@nstolpe if you can push your latest changes (even if the tests break) then I'll have a look

nstolpe commented:
@michalochman sorry I wasn't able to get to this sooner, I haven't had much time for coding outside of work recently.

I've pulled in your changes from 132 and implemented most of your suggestions. There are a few areas I'm a bit stuck on, if you could take a look and give some pointers.

1. I still haven't been able to get the correct dependencies for `useLayoutEffect` inside of `usePixiAppCreator`, or figured out how to set the returned `app` value without using `useState`

2. I've added the `createPixiApplication` function to `Stage` and used it in `createStageClass` and `usePixiAppCreator`. Functionally, this is working fine, but I haven't had any luck mocking it for use in the test (`appTestHook` is still in use for testing the function `Stage`).

3. I think there may be an inconsistency for the linting rules when linting locally via yarn/npm and when Code Climate linting is performed after pushing to GH. It's requiring [this formatting](https://github.com/michalochman/react-pixi-fiber/pull/127/commits/35ad8b4e7977129d4c6c77807eb3510dcd9a8159#diff-95b96dc34a5bf5b035b35940dd2d2d60R16) for `usePixiTicker` to pass after adding `fn` as a `useEffect` dependency, but running `yarn lint` locally fails unless it's formatted like this:

    ```javascript
    export function usePixiTicker(fn) {
      const app = usePixiApp();

      useEffect(() => {
        app.ticker.add(fn);

        return () => {
          app.ticker.remove(fn);
        };
      }, [fn]);
    }
    ``` 

4. After incorporating the changes from 132, the tests for hooks are passing, but I'm seeing this error and haven't figured out how to get rid of it:

    ```bash
    console.error node_modules/react-test-renderer/cjs/react-test-renderer.development.js:101
      Warning: Detected multiple renderers concurrently rendering the same context provider. This is currently unsupported.
    ```
    It's caused by the `usePixiTicker` test.

michalochman commented:
Thanks @nstolpe, much appreciated. I'll have a look at it this week.

michalochman commented:
@nstolpe I have not yet looked at 1., but 2. I have nearly resolved the test issues related to removal of `appTestHook`.

3. and 4. are minor, I'll deal with them later.

michalochman commented:
@nstolpe – tests updated.

1. I have added react hooks rules for eslint – you can now see what dependencies are needed by running eslint.
2. This is now resolved.
3. I have updated codeclimate and eslint config and it is now resolved.
4. You can ignore this for now.
5. This one is new – there are circular dependencies now that we should resolve soon: `Circular dependency: src/Stage.js -> src/hooks.js -> src/Stage.js` (reported by rollup)

nstolpe commented:
@michalochman 1 and [5](https://github.com/michalochman/react-pixi-fiber/pull/127/files#diff-12) fixed (added back stageProps).

5 is pretty straightforward.

1 is passing all tests and working with my test app. Beyond the dependencies, I had to add [this line](https://github.com/michalochman/react-pixi-fiber/pull/127/files#diff-afa643403bd6213fb932664673d00e3dR72) to `Stage.js`. Without that check, I would see the error below: 

![image](https://user-images.githubusercontent.com/1139918/64059842-dc67d700-cb91-11e9-8c41-81b3d1951239.png)

I did some debugging and `app` seemed to be defined, but had a bunch of null attributes, including `stage`. Just using `if (!app || !app.stage) return;` seems a little easy to me, but everything seems to be working fine.

michalochman commented:
Great @nstolpe! I think this is fine, but I'll try to test it.

Btw, scheduler is now available in `0.15.0`, I think we can try updating to see if the issue from `0.14.0` is gone and to fix #142.

RossKinsella commented:
Is there anything I can do to help with this?

nstolpe commented:
@michalochman I've updated scheduler, and the tests and build are working fine locally. CircleCI was failing with this error at install:

![image](https://user-images.githubusercontent.com/1139918/64574614-2abd6880-d33e-11e9-8295-da34f7d5d2eb.png)

I solved it by removing the dependency for scheduler 0.13.6 from `yarn.lock` and reinstalling, which [put the three scheduler dependencies in sequential order](https://github.com/michalochman/react-pixi-fiber/pull/127/commits/27378fede22ab96f51486867a6f60e866dc5d0bf) (0.13.1, 0.13.6, 0.15.0). Seems odd, but probably a yarn issue.

@RossKinsella if you'd like to pull down and test a build of the branch, you could see if it resolves your issue, or if there are any other issues.

RossKinsella commented:
Thanks @nstolpe, I'll take a look this evening!

RossKinsella commented:
Should we move scheduler into the peer dependencies? react-dom is dependent on it

nstolpe commented:
@michalochman awesome! Glad I could help out. Let me know how that testing goes, I'll hopefully be getting some work done on my project this week and will use this branch to see if I can spot anything either.

michalochman commented:
@RossKinsella `scheduler` is a direct dependency of this library. Moving it to `peerDependencies` would make it a requirement to users of this library to also install `scheduler` individually. As a user, all you should need to do is to install `react-pixi-fiber`.

RossKinsella commented:
@nstolpe I brought the branch into my project, which is a create-react-app managed one

I'm getting the error:
`Hooks can only be called inside the body of a function component. (https://fb.me/react-invalid-hook-call)`

with the following example:
```
// App.tsx
import React from 'react'
import { Stage } from 'react-pixi-fiber'

const App = () => <Stage width={600} height={400} />
```

## Stack trace:
invariant
src/my-awesome-app/node_modules/react-pixi-fiber/node_modules/react/cjs/react.development.js:85

resolveDispatcher
src/my-awesome-app/node_modules/react-pixi-fiber/node_modules/react/cjs/react.development.js:1470

useRef
src/my-awesome-app/node_modules/react-pixi-fiber/node_modules/react/cjs/react.development.js:1504

usePixiAppCreator
src/my-awesome-app/node_modules/react-pixi-fiber/cjs/react-pixi-fiber.development.js:14844

Stage
src/my-awesome-app/node_modules/react-pixi-fiber/cjs/react-pixi-fiber.development.js:14923

renderWithHooks
src/my-awesome-app/node_modules/react-dom/cjs/react-dom.development.js:15821

michalochman commented:
@nstolpe can you please verify if react-pixi-fiber now works with react version before 19.6.0? Updating reconciler used to break compatibility with earlier versions of react. No need to act on any issues that may happen, just worth knowing what the effect of the update is so I can start drafting some react compatibility strategy.

RossKinsella commented:
@nstolpe `useEffect` is working as expected now, thanks!

I have noticed that `useContext` is not working within a `<Stage>`, I have prepared an example where you can see this:
https://github.com/RossKinsella/react-pixi-hooks

Take note of:
https://github.com/RossKinsella/react-pixi-hooks/blob/master/src/MyContextConsumer.js
- This component is outside of the Stage and useContext works as expected

https://github.com/RossKinsella/react-pixi-hooks/blob/master/src/MyComponent.js
- This component is inside of the Stage and useContext is not working as expected

Is getting `useContext` working within the scope of this PR or should we open a new issue for this?

nstolpe commented:
@michalochman I've verified that react-pixi-fiber works with react and react-dom from 16.6.0 onward. I also tested with a build from master and also did not work on versions before 16.6.0. I've updated the package files for the main project and the examples.

I've also added a check to prevent errors from the hooks example when using an older version of react and some more examples of hooks.

aestrro commented:
Just wondering what's the hold up here? Any further support needed?

michalochman commented:
@aestrro the only hold up was lack of time for proper testing from my side.

nstolpe commented:
@michalochman I've been working off this branch w/ my project, and haven't encountered any issues yet. But lemme know if you do and I'll try and get them fixed quickly.

michalochman commented:
@nstolpe ok, I'll release it then and test it on my own anyway later

michalochman commented:
Released as 0.10.0
